fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,111 @@ where
         Ok((state, res))
     }
 }
+
+impl<DB> FvmMessageInterpreter<DB>
+where
+    DB: Blockstore + 'static + Send + Sync + Clone,
+{
+    fn estimate_gassed_msg(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: &mut Message,
+    ) -> anyhow::Result<Option<GasEstimate>> {
+        // Setting BlockGasLimit as initial limit for gas estimation
+        msg.gas_limit = BLOCK_GAS_LIMIT;
+
+        // estimate the gas limit and assign it to the message
+        // do not reuse the cache
+        let ret = state.call(msg.clone(), false)?;
+        if !ret.msg_receipt.exit_code.is_success() {
+            // if the message fail we can't estimate the gas.
+            return Ok(Some(GasEstimate {
+                exit_code: ret.msg_receipt.exit_code,
+                info: ret.failure_info.map(|x| x.to_string()).unwrap_or_default(),
+                gas_limit: 0,
```

Comment:
Should this be `gas_used`? In the [bottom](https://github.com/consensus-shipyard/fendermint/pull/163/files#diff-5be56262407092392424776bca11de06a243db9770516d04402d1f7d6961ab6bR194) it is, even in the error case.

Reply:
I used this to make explicit that the message errored and we shouldn't use the gas estimation, but no strong opinion, we can return the `gas_used` even if it failed.